<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>error</title>
  </head>
  <body>
    <button onclick="normalError()">normal error</button>
    <button onclick="promiseError()">promise error</button>
    <button onclick="promiseErrorInAsyncFunction()">promise error in try cache statement block</button>
    <script>
      function normalError() {
        console.log(a);
      }

      function promiseError() {
        new Promise((resolve, reject) => {
          setTimeout(() => {
            reject(b);
          }, 500);
        });
      }

      async function promiseErrorInAsyncFunction() {
        try {
          var res = await new Promise((resolve, reject) => {
            setTimeout(() => {
              reject('cancle');
            }, 500);
          });
        } finally {
        }
      }
      // DOM0级错误全局处理
      // window.onerror = function(message, source, lineno, colno, error) {
      //   console.log(message);
      // };

      // DOM2级错误全局处理
      window.addEventListener('error', function (event) {
        console.log('error: ' + event.message);
        event.preventDefault();
      });

      // Promise错误全局处理
      window.addEventListener('unhandledrejection', function (event) {
        console.log('unhandledrejection: ' + event.reason);
        event.preventDefault();
      });
    </script>
  </body>
</html>
